### PR TITLE
Implement a table-like shortcut to rpm macros in Lua

### PR DESCRIPTION
**Pull request by pmatilai**
Add rpm macro context as a global table-like entity named 'macros' for
a more Lua-native experience with rpm macros.

Support basic indexing syntaxes (macros[k] and macros.k) for access, define
and undefine operations. Undefined macros return nil here and assigning
to nil will undefine (pop) the macro.

As a specialty, parametric macros are returned as native callable
variadic Lua functions. A string argument is passed to expand as is,
but if arguments are passed as a table, eg `r = macros.foo({1, 2, 3})`,
they are passed entirely as-is.

The macro context pointer in the userdata is not consistently used for
all operations here, but then all the macro operations in Lua are hardwired
to the global context anyway so it doesn't matter in practise. Properly
passing the context around in all cases is left for later commits.

GH is not letting me reopen #1398 :unamused: so opening a new one to pick up where we left off there, now that @mlschroe did all the heavy lifting :smile:

### CODE REVIEW

File: rpmio/rpmlua.c [hunk 1]
```diff
@@ -1034,6 +1034,104 @@ static const luaL_Reg fd_m[] = {
...
+
+static int mc_newindex(lua_State *L)
+{
+    rpmMacroContext *mc = checkmc(L, 1);
+    const char *name = luaL_checkstring(L, 2);
+    if (lua_isnil(L, 3)) {
+	if (rpmPopMacro(*mc, name))
+	    luaL_error(L, "error undefining macro %s", name);
+    } else {
+	const char *body = luaL_checkstring(L, 3);
```
Comment (DemiMarie) [inline rpmio/rpmlua.c:1112]: We need to check for embedded NUL characters.

Comment (pmatilai) [reply rpmio/rpmlua.c:1112]: To what benefit? All that happens is the string will get terminated, I think people are likely to notice anyway...

Comment (DemiMarie) [reply rpmio/rpmlua.c:1112]: Throwing a useful error rather than a silent failure.

Comment (pmatilai) [reply rpmio/rpmlua.c:1112]: There's no actual error in that happening, one could even consider it a feature that strings are \0-terminated everywhere in rpm.

**Comment by pmatilai** [comment]
Dusted this off and made macros.define() and macros.global() work.

I think it's time to get this in. There are some quirks and inconsistencies present largely due to how convoluted this all is in the macro engine itself, but fixing all that is hard while there are no users for the code - ie this thing here. So unless there are some glaring bugs left, lets just get this merged so we can move forward with other stuff.

File: rpmio/rpmlua.c (hunk 1 above)
Comment (DemiMarie) [reply rpmio/rpmlua.c:1112]: Would it be possible to document this, and maybe give a warning?  It isn’t something I find intuitive.

Comment (pmatilai) [reply rpmio/rpmlua.c:1112]: We don't check for invalid characters in macros anywhere else, we're not going to do that here either.

**Comment by pmatilai** [comment]
Realized that documentation is missing, this has gone on so long now that a whole documentation project was started in the meanwhile...

**Comment by pmatilai** [comment]
Since there are no further comments...

**Pull request merged**